kernel/mem_domain.c

```diff
@@ -0,0 +1,304 @@
+/*
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <init.h>
+#include <kernel.h>
+#include <kernel_structs.h>
+#if defined(CONFIG_ARM_CORE_MPU)
+#include <arch/arm/cortex_m/mpu/arm_core_mpu.h>
+#endif
+
+
+static u8_t max_partitions;
+
+
+static inline u8_t arch_get_max_partitions(void)
+{
+#if defined(CONFIG_ARM_CORE_MPU)
+	return arm_core_mpu_get_max_domain_partition_regions();
+#else
+	return CONFIG_MAX_DOMAIN_PARTITIONS;
+#endif
+}
+
+static inline void arch_configure_mem_domain(struct k_thread *thread)
+{
+#if defined(CONFIG_ARM_CORE_MPU)
+	configure_mpu_mem_domain(thread);
+#else
+	x86_load_mmu_mem_domain(thread);
+#endif
+}
+
+static inline void arch_configure_mem_partition(struct k_thread *thread,
+						u32_t part_index)
+{
+#if defined(CONFIG_ARM_CORE_MPU)
+	configure_mpu_mem_partition(thread, part_index);
+#else
+	x86_load_mmu_mem_partition(thread, part_index);
+#endif
+
+}
+
+static inline void arch_disable_mem_partition(struct k_thread *thread,
+					      u32_t part_index)
+{
+#if defined(CONFIG_ARM_CORE_MPU)
+	disable_mpu_mem_partition(thread, part_index);
+#else
+	 x86_destroy_mem_partition(thread, part_index);
+#endif
+
+}
+
+static void arch_disable_all_mem_partitions(struct k_thread *thread)
+{
+#if defined(CONFIG_ARM_CORE_MPU)
+	disable_mpu_all_mem_partitions(thread);
+#else
+	x86_destroy_all_mem_partitions(thread);
+#endif
+}
+
+
+int k_mem_domain_init(struct k_mem_domain *domain, const char *name,
+		      u32_t num_parts, struct k_mem_partition *parts[])
+{
+	unsigned int key;
+
+	if (!domain || !name || (num_parts && !parts)) {
```

Comment:
Could these invariants be moved to assertions? If so, I'd move them to different assertions, so that, when they're logged, we know what argument was invalid.  Same for the test below and other functions.

Reply:
this is part of Chunlin's patch, I gave similar feedback in #814